### PR TITLE
eliminate broadcast_in_dim + transpose to broadcast_in_dim

### PR DESCRIPTION
**Pull request by qingyunqu**
transpose(broadcast_in_dim(X)) => broadcast_in_dim(X)

### CODE REVIEW

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 1]
```diff
@@ -4849,9 +4849,36 @@ static LogicalResult EliminateRedundantTranspse(TransposeOp op,
...
+  if (!broadcast_in_dim_op) {
+    return failure();
+  }
+  auto broadcast_dimensions = broadcast_in_dim_op.broadcast_dimensions();
+  auto permutation = op.permutation();
+  SmallVector<int64_t> new_broadcast_dimensions;
+  for (auto dimension : broadcast_dimensions) {
+    int64_t index = 0;
+    for (auto p : permutation) {
+      if (p == dimension) {
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:4865]: Can you make the type explicit for the 4 auto above please?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 2]
```diff
@@ -4849,9 +4849,36 @@ static LogicalResult EliminateRedundantTranspse(TransposeOp op,
...
+        new_broadcast_dimensions.push_back(index);
+        break;
+      }
+      index++;
+    }
+  }
+  rewriter.replaceOpWithNewOp<BroadcastInDimOp>(
+      op, op.getResult().getType(), broadcast_in_dim_op.operand(),
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:4873]:
```suggestion
      op, op.getResultType(), broadcast_in_dim_op.operand(),
```
I think `op.getResultType()` should work here

Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc]: There is no `getResultType()` of `TransposeOp`, I use `op->getResultTypes()` instead.

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 1 above)
Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:4866]: Done, thanks.

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 3]
```diff
@@ -4849,9 +4849,37 @@ static LogicalResult EliminateRedundantTranspse(TransposeOp op,
...
+  if (!broadcast_in_dim_op) {
+    return failure();
+  }
+  DenseIntElementsAttr broadcast_dimensions =
+      broadcast_in_dim_op.broadcast_dimensions();
+  DenseIntElementsAttr permutation = op.permutation();
+  SmallVector<int64_t> new_broadcast_dimensions;
+  for (APInt dimension : broadcast_dimensions) {
+    int64_t index = 0;
+    for (APInt p : permutation) {
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:4865]:
```suggestion
  for (auto dimension : broadcast_dimensions.getValues<int64_t>()) {
    int64_t index = 0;
    for (autp p : permutation.getValues<int64_t>()) {
```

And that's a good example of why `auto` is obfuscating the code :)

**Review by joker-eph — COMMENTED** [review]
LG, would the opposite fusion work as well? broadcast_in_dim(transpose(x)) -> broadcast_in_dim(x) ?

(no need to bundle it in this PR)

**Review by joker-eph — APPROVED** [review]

**Comment by qingyunqu** [comment]
I fix the typo, please approve again.  @joker-eph

**Review by joker-eph — APPROVED** [review]

**Comment by joker-eph** [comment]
> LG, would the opposite fusion work as well? broadcast_in_dim(transpose(x)) -> broadcast_in_dim(x) ?

You didn't answer me here, WDYT?

**Comment by qingyunqu** [comment]
> > LG, would the opposite fusion work as well? broadcast_in_dim(transpose(x)) -> broadcast_in_dim(x) ?
> 
> You didn't answer me here, WDYT?

Yes, I think `broadcast_in_dim(transpose(x)) -> broadcast_in_dim(x)` would work as well.  
But I think there maybe some pattern like `broadcast_in_dim(transpose(constant))`, should this pattern be simplified `broadcast_in_dim(transposed_constant)` or `broadcast_in_dim(constant)` ?

**Comment by joker-eph** [comment]
Right now this PR seems to be failing `//tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/canonicalize:transpose.mlir.test`:

``
tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/canonicalize/transpose.mlir:51:62: error: undefined variable: ARG
 // CHECK: [[RET:%[a-zA-Z0-9]+]] = "mhlo.broadcast_in_dim"([[ARG]])
                                                             ^
```

File: tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/canonicalize/transpose.mlir [hunk 4]
```diff
@@ -40,3 +40,42 @@ func @eliminate_redundant_transpose(%arg : tensor<3x4x16x2xf32>) -> tensor<3x2x1
   // CHECK-NEXT: return [[RET]]
   return %1 : tensor<3x2x16x4xf32>
 }
+
+// -----
+
+// CHECK-LABEL: func @broadcast_transpose
+// CHECK-SMAE: [[ARG:%[a-zA-Z0-9]+]]
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/canonicalize/transpose.mlir:47]:
```suggestion
// CHECK-SAME: [[ARG:%[a-zA-Z0-9]+]]
```

File: tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/canonicalize/transpose.mlir [hunk 5]
```diff
@@ -40,3 +40,42 @@ func @eliminate_redundant_transpose(%arg : tensor<3x4x16x2xf32>) -> tensor<3x2x1
...
+    // CHECK: [[RET:%[a-zA-Z0-9]+]] = "mhlo.broadcast_in_dim"([[ARG]])
+    // CHECK-SAME: dense<1>
+    // CHECK-NEXT: return [[RET]]
+    return %1 : tensor<5x64x31x95xf32>
+}
+
+// -----
+
+// CHECK-LABEL: func @broadcast_transpose_non_dim
+// CHECK-SMAE: [[ARG:%[a-zA-Z0-9]+]]
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/canonicalize/transpose.mlir:60]:
```suggestion
// CHECK-SAME: [[ARG:%[a-zA-Z0-9]+]]
```

File: tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/canonicalize/transpose.mlir [hunk 6]
```diff
@@ -40,3 +40,42 @@ func @eliminate_redundant_transpose(%arg : tensor<3x4x16x2xf32>) -> tensor<3x2x1
...
+    // CHECK: [[RET:%[a-zA-Z0-9]+]] = "mhlo.broadcast_in_dim"([[ARG]])
+    // CHECK-SAME: dense<>
+    // CHECK-NEXT: return [[RET]]
+    return %1 : tensor<5x64x31x95xf32>
+}
+
+// -----
+
+// CHECK-LABEL: func @broadcast_transpose_multi_dim
+// CHECK-SMAE: [[ARG:%[a-zA-Z0-9]+]]
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/canonicalize/transpose.mlir:73]:
```suggestion
// CHECK-SAME: [[ARG:%[a-zA-Z0-9]+]]
```

**Review by joker-eph — APPROVED** [review]

**Comment by joker-eph** [comment]
> But I think there maybe some pattern like broadcast_in_dim(transpose(constant)), should this pattern be simplified broadcast_in_dim(transposed_constant) or broadcast_in_dim(constant) ?

Seems better to not touch the constant in general if we don't need to, but that's hard to control: the folding of `transpose(constant)` is likely to happen independently first.

**Comment by joker-eph** [comment]
Sorry for the delay: this change exposed a few bugs in other places and I had to fix them so that all of our testing would pass! :)

**Pull request merged**